ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/KSPlanInfoExtractor.java

```diff
@@ -0,0 +1,191 @@
+  }
+
+  @Override
+  public KSPlanInfo visitStreamSelectKey(final StreamSelectKeyV1 streamSelectKey) {
+    return visitRepartitionStep(streamSelectKey);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitStreamSelectKey(final StreamSelectKey<K> streamSelectKey) {
+    return visitRepartitionStep(streamSelectKey);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitStreamSink(final StreamSink<K> streamSink) {
+    return visitSingleSourceStep(streamSink);
+  }
+
+  @Override
+  public KSPlanInfo visitStreamSource(final StreamSource streamSource) {
+    return visitSourceStep(streamSource);
+  }
+
+  @Override
+  public KSPlanInfo visitWindowedStreamSource(final WindowedStreamSource windowedStreamSource) {
+    return visitSourceStep(windowedStreamSource);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitStreamStreamJoin(final StreamStreamJoin<K> streamStreamJoin) {
+    return visitJoinStep(streamStreamJoin);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitStreamTableJoin(final StreamTableJoin<K> streamTableJoin) {
+    return visitJoinStep(streamTableJoin);
+  }
+
+  @Override
+  public KSPlanInfo visitTableSource(final TableSource tableSource) {
+    return visitSourceStep(tableSource);
+  }
+
+  @Override
+  public KSPlanInfo visitWindowedTableSource(final WindowedTableSource windowedTableSource) {
+    return visitSourceStep(windowedTableSource);
+  }
+
+  @Override
+  public KSPlanInfo visitStreamWindowedAggregate(
+      final StreamWindowedAggregate streamWindowedAggregate
+  ) {
+    return visitSingleSourceStep(streamWindowedAggregate);
+  }
+
+  @Override
+  public KSPlanInfo visitTableAggregate(final TableAggregate tableAggregate) {
+    return visitSingleSourceStep(tableAggregate);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitTableFilter(final TableFilter<K> tableFilter) {
+    return visitSingleSourceStep(tableFilter);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitTableGroupBy(final TableGroupBy<K> tableGroupBy) {
+    return visitSingleSourceStep(tableGroupBy);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitTableSelect(final TableSelect<K> tableSelect) {
+    return visitSingleSourceStep(tableSelect);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitTableSelectKey(final TableSelectKey<K> tableSelectKey) {
+    return visitRepartitionStep(tableSelectKey);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitTableSink(final TableSink<K> tableSink) {
+    return visitSingleSourceStep(tableSink);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitTableSuppress(final TableSuppress<K> tableSuppress) {
+    return visitSingleSourceStep(tableSuppress);
+  }
+
+  @Override
+  public <K> KSPlanInfo visitTableTableJoin(final TableTableJoin<K> tableTableJoin) {
+    return visitJoinStep(tableTableJoin);
+  }
+
+  private KSPlanInfo visitSourceStep(final ExecutionStep<?> step) {
+    return new KSPlanInfo(step);
+  }
+
+  private KSPlanInfo visitRepartitionStep(final ExecutionStep<?> step) {
```

Comment:
Today, we only ever repartition tables to enable joins, and it is not possible to repartition a table after a table operation that requires materialization (e.g., joining on a table that is the result of grouping by another table, in the same query) so it is correct to only track repartitions when deciding whether to force materialization at the table source or not. However, if in the future we allow table repartitions after table operations that require materialization, then this logic will have to be updated to account for that. Does it make sense to future-proof this and add in the more complicated logic now, or would that be overcomplicating things?

Reply:
I'm going to merge this PR without future-proofing as there are other PRs blocked behind this.

@agavra do you think it makes sense to try to future-proof this logic in a follow-up PR? It should be a small change. The reason I'm hesitant is because it requires making strong assumptions about which steps require materialization at the Kafka Streams level, and it feels weird to bake in these assumptions as they could in theory evolve in the future.

Reply:
The benefit of baking in the assumptions is that if they evolve we'll be conscious about it. I'm ok either way since we have pretty good test coverage here; we will catch in QTT if things change unexpectedly.

Reply:
Discussed (thoroughly 😂 ) offline: my concern is that if we support a new type of query that allows table aggregations prior to joins (in the same query), we may forget to update the logic in PlanInfoExtractor at that time, exposing us to the Avro schema name bug that the forced materialization at table sources tries to avoid. This would not be caught by existing QTTs as it would be for a new type of query. Summary of our discussion including how to proceed is here: https://github.com/confluentinc/ksql/issues/6750